libs/plugins/stream-server/src/lib/import/stream.ts

```diff
@@ -0,0 +1,83 @@
+import { Resource, ContributionType } from '@flogo-web/core';
+import {
+  StreamData,
+  StreamMetadata,
+  InternalStage,
+  StreamActionSettings,
+  StreamResourceModel,
+} from '@flogo-web/plugins/stream-core';
+import { ResourceImportContext, ValidationError } from '@flogo-web/lib-server/core';
+
+import { STREAM_POINTER } from '../constants';
+import { makeResourceValidator } from './make-resource-validator';
+
+export function importStreamResource(
+  resource: Resource<StreamData>,
+  context: ResourceImportContext
+): Resource<StreamData> {
+  const validate = makeResourceValidator(
+    Array.from(context.contributions.keys()),
+    context.importsRefAgent
+  );
+  const errors = validate(resource);
+  if (errors) {
+    throw new ValidationError('Stream data validation errors', errors);
+  }
+  return {
+    ...resource,
+    metadata: extractMetadata(resource, context),
+    data: {
+      stages: extractStages(resource, context),
+    },
+  };
+}
+
+function extractMetadata(
+  resource: Resource,
+  { actionsManager, normalizedResourceIds }: ResourceImportContext
+): StreamMetadata {
+  const metadata = { ...resource.metadata };
+  const originalResourceId = getOriginalId(normalizedResourceIds, resource.id);
+  const actionSettings = actionsManager.getSettingsForResourceId(
+    originalResourceId,
+    STREAM_POINTER
+  ) as StreamActionSettings;
+  return {
+    input: metadata.input || [],
+    output: metadata.output || [],
+    groupBy: actionSettings && actionSettings.groupBy,
+  };
+}
+
+function extractStages(
+  resource: Resource<StreamData>,
+  context: ResourceImportContext
+): InternalStage[] {
+  const getStageReference = (alias: string) => {
+    return context.importsRefAgent.getPackageRef(ContributionType.Activity, alias);
+  };
+  const rawStages = (resource.data && resource.data.stages) || [];
+  return rawStages.map((stage, index) => {
+    stage.ref = getStageReference(stage.ref);
+    return formatStage(stage, index);
+  });
+}
+
+function getOriginalId(source: Map<string, string>, newId: string): string {
+  return Array.from(source.entries()).find(([, resourceId]) => resourceId === newId)[0];
```

Comment:
how about using an for..of and an early return to avoid converting the whole map to an array?
```
for (let [oldResourceId, newResourceId] of source) {
    if (newResourceId === newId) { 
      return oldResourceId; 
    }
  }
```

Also, should we provided in the context the reverse map of `normalizedResourceIds`? This way this conversion at the plugin level can be avoided.

Reply:
I think providing a reverse map is a better idea as it might be needed for other plugins which work in co-ordination with actions. I have created a ticket #1177 to implement this change